src/Language/Spectacle/Syntax/Logic.hs

```diff
@@ -110,51 +109,45 @@ runLogic ::
   Lang ctx (Logic ': effs) a ->
   Lang ctx effs a
 runLogic = \case
-  Pure x -> Pure x
-  Yield (Op op) k -> case decomposeOp op of
-    Left other -> Yield (Op other) (runLogic . k)
+  Pure x -> pure x
+  Op op k -> case decomposeOp op of
+    Left other -> Op other (runLogic . k)
     Right bottom -> absurd (coerce bottom)
-  Yield (Scoped scoped loom) k -> case decomposeS scoped of
-    Left other -> Yield (Scoped other loom') k'
-    Right (Forall xs p) ->
-      k' =<< runLoom loom' do
-        b <- fmap and (mapM p xs)
-        unless b (throwE (QuantifierException ForallViolated))
-        oneOf xs
-    Right (Exists xs p) ->
-      k' =<< runLoom loom' do
-        xs' <- filterM p xs
-        when (null xs') (throwE (QuantifierException ExistsViolated))
-        oneOf xs'
-    Right (Complement m) -> runLoom loom' (rewriteComplement m) >>= k'
-    Right (Conjunct m n) -> runLoom loom' (liftA2 (&&) m n) >>= k'
-    Right (Disjunct m n) -> runLoom loom' (liftA2 (||) m n) >>= k'
+  Scoped scoped loom -> case decomposeS scoped of
+    Left other -> Scoped other loom'
+    Right (Forall xs p) -> runLoom loom' do
+      b <- fmap and (mapM p xs)
+      unless b (throwE (QuantifierException ForallViolated))
+      return b
+    Right (Exists xs p) -> runLoom loom' do
+      b <- fmap or (mapM p xs)
+      unless b (throwE (QuantifierException ExistsViolated))
+      return b
+    Right (Complement m) -> runLoom loom' (rewriteComplement m)
+    Right (Conjunct m n) -> runLoom loom' (liftA2 (&&) m n)
+    Right (Disjunct m n) -> runLoom loom' (m <|> n)
     where
       loom' = loom ~>~ hoist runLogic
 
-      k' = runLogic . k
-
 -- | Applies negation to logical operators.
 --
 -- @since 0.1.0.0
-rewriteComplement :: Member Logic effs => Lang ctx effs a -> Lang ctx effs a
+rewriteComplement :: Members Logic effs => Lang ctx effs a -> Lang ctx effs a
 rewriteComplement = \case
   Pure x -> pure x
-  Yield (Op op) k -> Yield (Op op) (rewriteComplement . k)
-  Yield (Scoped scoped loom) k -> case projectS scoped of
-    Nothing -> Yield (Scoped scoped loom') k'
-    Just (Forall xs p) -> runLoom loom (exists xs (fmap not . p)) >>= k'
-    Just (Exists xs p) -> runLoom loom (forall xs (fmap not . p)) >>= k'
-    Just (Complement m) -> runLoom loom' m >>= k'
+  Op op k -> Op op (rewriteComplement . k)
+  Scoped scoped loom -> case projectS scoped of
+    Nothing -> Scoped scoped loom'
+    Just (Forall xs p) -> runLoom loom (exists xs (fmap not . p))
+    Just (Exists xs p) -> runLoom loom (forall xs (fmap not . p))
+    Just (Complement m) -> runLoom loom' m
     Just (Conjunct m n) -> do
```

Comment:
DeMorgan's laws are well known enough that we _probably_ don't need to call out both the negation of a conjunction and disjunction. I had the stray thought though that it might be nice for some readers to name this and the following case after the law they're expressing. What do you think?